src/Ocelot/RateLimit/RateLimitCore.cs

```diff
@@ -0,0 +1,123 @@
+﻿using Ocelot.Configuration;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Ocelot.RateLimit
+{
+    public class RateLimitCore
+    {
+        private readonly IRateLimitCounterHandler _counterHandler;
+        private static readonly object _processLocker = new object();
+
+        public RateLimitCore(IRateLimitCounterHandler counterStore)
+        {
+            _counterHandler = counterStore;
+        }
+
+        public RateLimitCounter ProcessRequest(ClientRequestIdentity requestIdentity, RateLimitOptions option)
+        {
+            var counter = new RateLimitCounter
+            {
+                Timestamp = DateTime.UtcNow,
+                TotalRequests = 1
+            };
+            var rule = option.RateLimitRule;
+
+            var counterId = ComputeCounterKey(requestIdentity, option);
+
+            // serial reads and writes
+            lock (_processLocker)
+            {
+                var entry = _counterHandler.Get(counterId);
+                if (entry.HasValue)
+                {
+                    // entry has not expired
+                    if (entry.Value.Timestamp + rule.PeriodTimespan.Value >= DateTime.UtcNow)
```

Comment:
PeriodTimespan could be null here, maybe it should not be nullable or we could change to see if it is null before .Value